### PR TITLE
Implement groupby over index

### PR DESCRIPTION
**Pull request by phofl**
This addresses a couple of benchmark failures

### CODE REVIEW

File: dask_expr/_groupby.py [hunk 1]
```diff
@@ -188,6 +193,8 @@ def spec(self):
 
     @classmethod
     def chunk(cls, df, by=None, **kwargs):
+        if hasattr(by, "dtype"):
```
Comment (phofl) [inline dask_expr/_groupby.py:196]: Not too happy about this, open for suggestions

**Review by rjzamora — COMMENTED** [review]
I took a quick glance at this, and plan to look more carefully later today.

One immediate thought I have is whether there is any way to avoid adding groupby-specific attributes/properties (like "by") to the classes defined in `_reductions.py`. At a high level, I understand why you took this approach, but I'm still hopeful that we can keep the "general" classes in `_reductions` a bit cleaner. (though, a closer look may convince me otherwise).

**Comment by rjzamora** [comment]
>We have to pack this into args somehow because it can be an expression, open to other approaches. Not happy with this myself...

Right, that makes sense. However, it *may* make the most sense (in the long run) to add some kind of "base" `ApplyConcatApplyBy` and `ChunkBy` classes in `_groupby.py`, and to take the time to keep groupby details out of the general reduction classes.

**Comment by phofl** [comment]
I tried this, but it does a lot of copy-paste things that I liked even less. Ideally I agree with you though

**Comment by rjzamora** [comment]
Just leaving a note that this PR was superseded by https://github.com/dask-contrib/dask-expr/pull/271

**Pull request closed without merging**